atomkraft/reactor_room/reactor_room.py

```diff
@@ -0,0 +1,109 @@
+import ast
+from os import PathLike
+import os
+from typing import List
+import tomlkit
+from . import constants
+from . import utils
+from .step_functions_visitor import StepFunctionsVisitor
+from atomkraft.utils.project import project_root
+
+
+def check_reactor(trace: PathLike, keypath="action", reactor=None) -> bool:
+    """
+    returns true if each action appearing in `trace` is matched with a function in `reactor`
+    """
+    if reactor is None:
+        reactor = get_reactor()
+
+    with open(reactor) as f_reactor:
+        root_node = ast.parse(f_reactor.read())
+    v = StepFunctionsVisitor()
+    v.visit(root_node)
+    all_trace_actions = utils.get_all_trace_actions(trace, keypath)
+
+    if constants.ALL_ENCOMPASSING_STEP in v.step_functions:
+        return True
+    else:
+        return all_trace_actions.issubset(v.step_functions)
+
+
+def get_reactor() -> PathLike:
+    """
+    returns the path to the current reactor from the internal config
+    """
+    internal_config_file_path = os.path.join(
+        project_root(),
+        constants.ATOMKRAFT_INTERNAL_FOLDER,
+        constants.ATOMKRAFT_INTERNAL_CONFIG,
+    )
+    with open(internal_config_file_path) as config_f:
+        config_data = tomlkit.load(config_f)
+        return config_data[constants.REACTOR_CONFIG_KEY]
+
+
+def generate_reactor(
+    actions_list: List[str], variables_list: List[str], stub_file_path: PathLike = None
+) -> PathLike:
```

Comment:
I believe the API here needs to be changed from:

```python
def check_reactor(trace: PathLike, keypath="action", reactor=None) -> bool:

def get_reactor() -> PathLike

def generate_reactor(
    actions_list: List[str], variables_list: List[str], stub_file_path: PathLike = None
) -> PathLike
```

to

```python
def check_reactor(trace: PathLike, reactor: PathLike) -> bool:

def get_reactor() -> PathLike

def generate_reactor(
    actions_list: List[str], variables_list: List[str], keypath="action", stub_file_path: PathLike = None
) -> PathLike
```

The point is that the `keypath` argument should be considered as part of the reactor. The user, when they call `test trace`, which under the hood calls `get_reactor` and `check_reactor` should not provide the `keypath` argument each time. Thus, it should be moved to the `generate_reactor` function.

Reply:
I  agree for the workflow of `test trace`. But what about when the user wrote his own reactor and wants to check it against a trace?  

What I could do is to give this signature
`def check_reactor(trace: PathLike, keypath=None, reactor=None) -> bool:`, and if `keypath` is `None`, then to look into the config file (where I would store the `keypath` upon creating the reactor).

Reply:
resolved synchronously: `keypath` should be a part of the generated file (I misunderstood that part initially).